ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs

```diff
@@ -565,49 +577,335 @@ toPublicState PeerSelectionState { knownPeers
+
+-- | Internal function, used to pattern match with `PeerSelectionCountersHWC`
+peerSelectionCountersHWC :: PeerSelectionCounters -> PeerSelectionCounters
+peerSelectionCountersHWC PeerSelectionCounters {..} =
+    PeerSelectionCounters {
+      numberOfRootPeers,
+
+      numberOfKnownPeers                         = numberOfKnownPeers
+                                                 - numberOfEstablishedPeers,
+      numberOfColdPeersPromotions,
+      numberOfEstablishedPeers                   = numberOfEstablishedPeers
+                                                 - numberOfActivePeers,
+      numberOfWarmPeersDemotions,         
+      numberOfWarmPeersPromotions,         
+      numberOfActivePeers,
+      numberOfActivePeersDemotions,
+
+      numberOfKnownBigLedgerPeers                = numberOfKnownBigLedgerPeers
+                                                 - numberOfEstablishedBigLedgerPeers,
+      numberOfColdBigLedgerPeersPromotions,
+      numberOfEstablishedBigLedgerPeers          = numberOfEstablishedBigLedgerPeers
+                                                 - numberOfActiveBigLedgerPeers,
+      numberOfWarmBigLedgerPeersDemotions,
+      numberOfWarmBigLedgerPeersPromotions,
+      numberOfActiveBigLedgerPeers,
+      numberOfActiveBigLedgerPeersDemotions,
+
+      numberOfKnownLocalRootPeers                = numberOfKnownLocalRootPeers
+                                                 - numberOfEstablishedLocalRootPeers,
+      numberOfEstablishedLocalRootPeers          = numberOfEstablishedLocalRootPeers
+                                                 - numberOfActiveLocalRootPeers,
+      numberOfWarmLocalRootPeersPromotions,
+      numberOfActiveLocalRootPeers,
+      numberOfActiveLocalRootPeersDemotions,
+
+      localRoots,
+
+      numberOfKnownSharedPeers                   = numberOfKnownSharedPeers
+                                                 - numberOfEstablishedSharedPeers,
+      numberOfColdSharedPeersPromotions,
+      numberOfEstablishedSharedPeers             = numberOfEstablishedSharedPeers
+                                                 - numberOfActiveSharedPeers,
+      numberOfWarmSharedPeersDemotions,
+      numberOfWarmSharedPeersPromotions,
+      numberOfActiveSharedPeers,
+      numberOfActiveSharedPeersDemotions,
+
+      numberOfKnownBootstrapPeers                = numberOfKnownBootstrapPeers
+                                                 - numberOfEstablishedBootstrapPeers,
+      numberOfColdBootstrapPeersPromotions,
+      numberOfEstablishedBootstrapPeers          = numberOfEstablishedBootstrapPeers
+                                                 - numberOfActiveBootstrapPeers,
+      numberOfWarmBootstrapPeersDemotions,
+      numberOfWarmBootstrapPeersPromotions,
+      numberOfActiveBootstrapPeers,
+      numberOfActiveBootstrapPeersDemotions
+    }
+
+
 peerStateToCounters :: Ord peeraddr => PeerSelectionState peeraddr peerconn -> PeerSelectionCounters
```

Comment:
The `peerStateToCounters` function is simple but rather easy to make a sill mistake which would lead to wrong churn behaviour or misleading tracing output.